lib/utils/property-order.js

```diff
@@ -1,39 +1,162 @@
+'use strict';
+
 var ember = require('./ember');
 
 module.exports = {
+  determinePropertyType,
   reportUnorderedProperties: reportUnorderedProperties
 };
 
-function reportUnorderedProperties(node, context, toPropertyInfo) {
+const NAMES = {
+  'service': 'service injection',
+  'property': 'property',
+  'single-line-function': 'single-line function',
+  'multi-line-function': 'multi-line function',
+  'observer': 'observer',
+  'model': '"model" hook',
+  'lifecycle-hook': 'lifecycle hook',
+  'actions': 'actions hash',
+  'method': 'method',
+  'unknown': 'unknown property type',
+  'attribute': 'attribute',
+  'relationship': 'relationship',
+};
+
+function determinePropertyType(node, parentType) {
+  if (ember.isInjectedServiceProp(node.value)) {
+    return 'service';
+  }
+
+  if (parentType === 'component') {
+    if (ember.isComponentLifecycleHook(node)) {
+      return 'lifecycle-hook';
+    }
+  }
+
+  if (parentType === 'controller') {
+    if (ember.isControllerDefaultProp(node)) {
+      return 'inherited-property';
+    }
+  }
+
+  if (parentType === 'model') {
+    if (ember.isModule(node.value, 'attr', 'DS')) {
+      return 'attribute';
+    } else if (ember.isRelation(node)) {
+      return 'relationship';
+    }
+  }
+
+  if (parentType === 'route') {
+    if (ember.isRouteDefaultProp(node)) {
+      return 'inherited-property';
+    } else if (ember.isModelProp(node)) {
+      return 'model';
+    } else if (ember.isRouteDefaultMethod(node)) {
+      return 'lifecycle-hook';
+    }
+  }
+
+  if (ember.isObserverProp(node.value)) {
+    return 'observer';
+  }
+
+  if (parentType !== 'model' && ember.isActionsProp(node)) {
+    return 'actions';
+  }
+
+  if (ember.isSingleLineFn(node)) {
+    return 'single-line-function';
+  }
+
+  if (ember.isMultiLineFn(node)) {
+    return 'multi-line-function';
+  }
+
+  if (ember.isCustomProp(node)) {
+    return 'property';
+  }
+
+  if (ember.isFunctionExpression(node.value)) {
+    return 'method';
+  }
+
+  return 'unknown';
+}
+
+function getOrder(ORDER, type) {
+  for (var i = 0, len = ORDER.length; i < len; i++) {
```

Comment:
you can't use `forEach` here because there is no way to break out of a `forEach` before having iterated over the whole array. in this case though we can stop iterating once we've found what we're looking for.